### PR TITLE
Refactor `_apply_unitary_to_effect_` into `_apply_unitary_`

### PR DESCRIPTION
**Pull request by Strilanc**
- Combine arguments into an args object (for forwards compatibility and easier-to-remember signature)
- Add a `subspace_index` helper method to the args object
- `None` is now a valid value for indicating that the current value can't `_apply_unitary_` (equivalent to `NotImplemented`)
- Fixed a few type annotations along the way

### CODE REVIEW

File: cirq/protocols/apply_unitary.py [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+        The receiving object may also write its output over a new buffer
+        that it created, in which case that new array is returned.
+
+    Raises:
+        TypeError: `unitary_value` doesn't have a unitary effect and `default`
+            wasn't specified.
+    """
+
+    # Check if the specialized method is present.
+    getter = getattr(unitary_value, '_apply_unitary_', None)
```
Comment (viathor) [inline cirq/protocols/apply_unitary.py]: nit: `getter` is a poor name here (it suggests an existing value is accessed, in reality a new value is likely being computed), suggestion:

```python
apply = getattr(unitary_value, '_apply_unitary_', None)
if apply is not None:
    result = apply(args)
    ...
```

File: cirq/protocols/apply_unitary.py [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+            that the result is in `args.available_buffer` (and so what was
+            `args.available_buffer` will become `args.target_tensor` in the next
+            call, and vice versa).
+
+            The receiving object is also permitted to allocate a new
+            numpy.ndarray and return that as its result.
+        """
+
+
+def apply_unitary(unitary_value: Any,
```
Comment (viathor) [inline cirq/protocols/apply_unitary.py]: nit: possibly better name: `unitary`.

Comment (viathor) [reply cirq/protocols/apply_unitary.py]: Clarification: s/unitary_value/unitary/

File: cirq/protocols/apply_unitary.py [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+            unitary transformation or some other concept.
+        available_buffer: Pre-allocated workspace with the same shape and
+            dtype as the target tensor.
+        axes: Which axes the unitary effect is being applied to (e.g. the
+            qubits that the gate is operating on).
+    """
+
+    def __init__(self,
+                 target_tensor: np.ndarray,
+                 available_buffer: np.ndarray,
```
[the remark this answers is not in the document]
Comment (Strilanc) [reply cirq/protocols/apply_unitary.py]: Yes. Actually, `cirq.apply_unitary` should probably allow a calling style where you say `state: Union[int, np.ndarray]`. But it's a bit messy, so I deferred it.

File: cirq/protocols/apply_unitary.py (hunk 2 above)
Comment (Strilanc) [reply cirq/protocols/apply_unitary.py]: Acknowledged. `unitary` would sound like it was asking for the matrix, I think. Left as is.

File: cirq/protocols/apply_unitary.py (hunk 1 above)
Comment (Strilanc) [reply cirq/protocols/apply_unitary.py]: Renamed to `func`.

**Pull request merged**